packages/core/lib/use-puck-data.ts

```diff
@@ -0,0 +1,136 @@
+    if (action.type === "insert") {
+      const forward = () => _dispatch(action);
+      const rewind = () =>
+        _dispatch({
+          type: "remove",
+          index: action.destinationIndex,
+          zone: action.destinationZone,
+        });
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "reorder") {
+      const forward = () => _dispatch(action);
+      const rewind = () =>
+        _dispatch({
+          type: "reorder",
+          sourceIndex: action.destinationIndex,
+          destinationIndex: action.sourceIndex,
+          destinationZone: action.destinationZone,
+        });
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "move") {
+      const forward = () => _dispatch(action);
+      const rewind = () =>
+        _dispatch({
+          type: "move",
+          sourceIndex: action.destinationIndex,
+          sourceZone: action.destinationZone,
+          destinationIndex: action.sourceIndex,
+          destinationZone: action.destinationZone,
+        });
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "replace") {
+      const currentProps = selectedItem?.props || data.root;
+      const forward = () => _dispatch(action);
+      const rewind = () =>
+        _dispatch({
+          ...action,
+          data: { ...selectedItem, props: currentProps },
+        });
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "set") {
+      const currentProps = selectedItem?.props || data.root;
+      const forward = () => _dispatch(action);
+      const rewind = () =>
+        _dispatch({
+          ...action,
+          data: { root: currentProps as any },
+        });
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "duplicate") {
+      const forward = () => _dispatch(action);
+      const rewind = () => {
+        _dispatch({
+          type: "remove",
+          index: action.sourceIndex,
+          zone: action.sourceZone,
+        });
+        setItemSelector(null);
+      };
+
+      record({ forward, rewind });
+      forward();
+    } else if (action.type === "remove") {
+      const currentItem = getItem(
+        { index: action.index, zone: action.zone },
+        data
+      );
+      const forward = () => _dispatch(action);
+      const rewind = () => {
+        _dispatch({
+          type: "recovery",
+          destinationIndex: action.index,
+          destinationZone: action.zone,
+          data: currentItem,
+        });
+        setItemSelector({
+          zone: action.zone,
+          index: action.index,
+        });
+      };
+
+      record({ forward, rewind });
+      forward();
+    }
+  };
+
+  return {
+    data,
+    dispatch,
+    canForward,
+    canRewind,
+    rewind,
+    forward,
+    record,
+    itemSelector,
+    setItemSelector,
+    selectedItem,
+  };
+}
```

Comment:
Hm I'm a bit unsure about this particular file. Perhaps we would be be better off going with deep object diff library, and store the delta, rather than re-calling the dispatchers here.

You could use [deep-diff](https://www.npmjs.com/package/deep-diff) and the [applyChange API](https://github.com/flitbit/diff/tree/e2b5b86add402ae4373ab025ee7e9a98e3424023#changes), or try out [rfc6902](https://github.com/chbrown/rfc6902). I don't think it really matters which.

Example using `deepDiff` inside `reducer.ts`, without modifying existing reducer code to avoid polluting the state machine

```tsx

const storeHistory = (reducer: StateReducer) => {
  const history: any = [];

  return (data, action) => {
    const newData = reducer(data, action);

    const diff = deepDiff.diff(data, newData);

    history.push(diff);

    return newData;
  };
};

export const createReducer = ({ config }: { config: Config }): StateReducer =>
  storeHistory((data, action) => {
    if (action.type === "insert") {
      const emptyComponentData = {
        type: action.componentType,
        props: {
          ...(config.components[action.componentType].defaultProps || {}),
          id: generateId(action.componentType),
        },
      };
   ...
```

Reply:
Oh this idea looks really cool!
I'll come back using this ⛹️‍♂️

Reply:
Hi! @chrisvxd
I tried applying deep-diff. I still have a little considered about each responsibility, but it ready to review 🔥